### PR TITLE
Implement Size() and Len() to know the optimal size.

### PR DESCRIPTION
**Pull request by harshavardhana**

### CODE REVIEW

File: internal/backend/s3/s3.go [hunk 1]
```diff
@@ -215,6 +262,18 @@ func (be *Backend) Save(ctx context.Context, h restic.Handle, rd io.Reader) (err
...
+		rd = nopCloserFile{f}
+	} else if l, ok := rd.(lenner); ok {
+		debug.Log("reader is %#T, using nopCloserLenner{}", rd)
+		rd = nopCloserLenner{l}
+	} else {
+		debug.Log("reader is %#T, no specific workaround enabled", rd)
+	}
+
 	// prevent the HTTP client from closing a file
 	rd = ioutil.NopCloser(rd)
```
Comment (fd0) [inline internal/backend/s3/s3.go]: This needs to be removed in order for the workaround above to be active :)

**Review by fd0 — CHANGES_REQUESTED** [review]

File: internal/backend/s3/s3.go (hunk 1 above)
Comment (harshavardhana) [reply internal/backend/s3/s3.go]: oops yes..

**Comment by atomi** [comment]
Confirming that this fixes #1256    

Thanks @harshavardhana

**Comment by fd0** [comment]
For the record, here are the benchmarks:
```
benchmark                                                  old ns/op     new ns/op     delta
BenchmarkBackendMinio/BenchmarkLoadFile-4                  6828941       7045987       +3.18%
BenchmarkBackendMinio/BenchmarkLoadPartialFile-4           2426226       2396428       -1.23%
BenchmarkBackendMinio/BenchmarkLoadPartialFileOffset-4     2457739       2388229       -2.83%
BenchmarkBackendMinio/BenchmarkSave-4                      179000485     82541729      -53.89%
benchmark                                                  old MB/s     new MB/s     speedup
BenchmarkBackendMinio/BenchmarkLoadFile-4                  2457.09      2381.40      0.97x
BenchmarkBackendMinio/BenchmarkLoadPartialFile-4           1729.18      1750.68      1.01x
BenchmarkBackendMinio/BenchmarkLoadPartialFileOffset-4     1707.01      1756.69      1.03x
BenchmarkBackendMinio/BenchmarkSave-4                      93.74        203.28       2.17x
benchmark                                                  old allocs     new allocs     delta
BenchmarkBackendMinio/BenchmarkLoadFile-4                  233            233            +0.00%
BenchmarkBackendMinio/BenchmarkLoadPartialFile-4           235            234            -0.43%
BenchmarkBackendMinio/BenchmarkLoadPartialFileOffset-4     237            237            +0.00%
BenchmarkBackendMinio/BenchmarkSave-4                      14613          14028          -4.00%
benchmark                                                  old bytes     new bytes     delta
BenchmarkBackendMinio/BenchmarkLoadFile-4                  15916         15910         -0.04%
BenchmarkBackendMinio/BenchmarkLoadPartialFile-4           16042         15985         -0.36%
BenchmarkBackendMinio/BenchmarkLoadPartialFileOffset-4     16278         16265         -0.08%
BenchmarkBackendMinio/BenchmarkSave-4                      605555007     1495782       -99.75%
```

**Pull request merged**